File: MapGUI3/src/Model/Statements/closeRFile.java
package Model.Statements;

import Model.FileTable;
import Model.ProgramState;
import Model.SymbolTable;
import Model.expressions.Expression;

import java.io.BufferedReader;
import java.io.IOException;

public class closeRFile implements IStatement {
    private Expression id;

    public closeRFile(Expression id){ this.id=id;}

    public ProgramState execute(ProgramState ps){
        FileTable ft=ps.getFileTable();
        SymbolTable<String,Integer> symTable = ps.getSymTable();
        int res=id.evaluate(symTable,ps.getHeap());
        BufferedReader buff = ft.getBufferReader(res);
        try {
            buff.close();
            ft.remove(res);
        } catch (IOException e) {
            e.printStackTrace();
        }
        return ps;
    }

    public String toString(){
        return "closeRFile(" + id.toString()+")" ;
    }
}

File: MapA7/MapA7/Model/Statements/CloseRFile.cs
using MapA7.Model.Exceptions;
using MapA7.Model.Expressions;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MapA7.Model.Statements
{
    public class CloseRFile:IStatement
    {
        private IExpression expr;

        public CloseRFile(IExpression expr)
        {
            this.expr = expr;
        }

        public ProgramState Execute(ProgramState ps)
        {
            SymbolTable symTable = ps.GetSymbolTable();
            FileTable fileTable = ps.GetFileTalbe();
            int fd = expr.Evaluate(symTable);
            StreamReader sr = fileTable.GetValue(fd);
            if (sr == null) throw new CustomException("The file ins't opened!.");
            try
            {
                sr.Close();
                fileTable.Remove(fd);
            }
            catch (IOException)
            {
                throw new CustomException("Closing of the file resulted in an error.");
            }
            return ps;
        }

        public String Tostring()
        {
            return "CloseRFile(" + expr.Tostring() + ")";
        }
    }
}
